app/src/main/java/org/listenbrainz/android/service/ApiService.kt

```diff
@@ -0,0 +1,148 @@
+package org.listenbrainz.android.service
+
+import org.listenbrainz.android.model.CoverArt
+import org.listenbrainz.android.model.ListenBrainzExternalServices
+import org.listenbrainz.android.model.ListenSubmitBody
+import org.listenbrainz.android.model.Listens
+import org.listenbrainz.android.model.PinData
+import org.listenbrainz.android.model.PinnedRecording
+import org.listenbrainz.android.model.PostResponse
+import org.listenbrainz.android.model.RecommendationData
+import org.listenbrainz.android.model.Review
+import org.listenbrainz.android.model.SearchResult
+import org.listenbrainz.android.model.SimilarUserData
+import org.listenbrainz.android.model.SocialData
+import org.listenbrainz.android.model.SocialResponse
+import org.listenbrainz.android.model.TokenValidation
+import org.listenbrainz.android.model.feed.FeedData
+import org.listenbrainz.android.model.feed.FeedEvent
+import org.listenbrainz.android.model.feed.FeedEventDeletionData
+import org.listenbrainz.android.model.feed.FeedEventVisibilityData
+import org.listenbrainz.android.util.Constants.Headers.AUTHORIZATION
+import retrofit2.Response
+import retrofit2.http.Body
+import retrofit2.http.GET
+import retrofit2.http.Header
+import retrofit2.http.POST
+import retrofit2.http.Path
+import retrofit2.http.Query
+
+interface ApiService {
```

Comment:
Why fit everything into one single service? I don't think this is a good thing to do. Please state your reasons to do so, otherwise revert

Reply:
How are you measuring the performance differences? We can't just go by words because theoretically speaking, "It's accurate that creating a Retrofit instance is relatively expensive in terms of performance, as it involves setting up a series of internal components, parsing annotations, and often establishing network configurations. However, the term "heavy" is relative. In most scenarios, it's a one-time cost that doesn't noticeably affect the end-user experience, especially if done during the app's initialization phase." 

That being said, I do not want the profile/dashboard services in a single file with feed services. It just doesn't make sense to pollute the codebase by merging everything into a single file. Keep those separate. If the app performance has not taken a huge toll and we have numbers to prove it, we should make sure we don't end up confusing the dev codebase for it.

Reply:
The different base urls reason is invalid btw.

Reply:
Well for once, LB website does it the same way, and having more Singletons have actual performance impact. For the Creation part, the heavy retrofit instances do impact the first time launch (crucial) but also take up chunks of memory without serving no actual use. These different retrofit services would serve us only if we had the microservices type architecture where codebase and endpoint is different for each service.

I'm just stating facts provided by the actual library docs and most of the community agrees.

Reply:
Also it isn't that confusing tbh. Its more precise such that a dev will automatically know where to look an API call when they need it.

Reply:
We have added less than half of the endpoints we need to add for stuff like taste, recommendations, playlists, etc. We can't have everything in one file. I would go for better code abstraction over meaningless performance improvement here.

Reply:
You think following a comment-based approach is better in code, but I disagree. Your code should speak for itself and trust me devs hardly read in-code documentation. Any file with too much code is just a pain to deal with. That is why we always like to prefer break things into components wherever possible. Also, don't take the LB website to the perfect example of a codebase. It is still nascent.

Reply:
Its okay tbh. Sometimes when the apps get big, abstractions only make things difficult. And the performance improvement is not negligible. If we keep this up, its going to bite us in future for sure.

Also, the interface is barely 160 lines long. People are usually fine with 1000+ loc as long as ctrl+f is there.

Reply:
@07jasjeet are you familiar with SOLID principles? Please read through those once. I would like us to follow only those and avoid conflicts. Also, please set up a performance monitor like Android profiler to suit arguments with numbers from next time. I am not trying to make the app less performant but being a developer, having a single ApiService makes me want to drive away from the codebase as far as possible, hence the resistance.

Reply:
Also, sacrificing app performance and memory usage for Developer experience is kinda absurd to me. If people can deal with it on ListenBrainz Server idk why we should soften things up.

Reply:
We don't need to force principles it'll make things more confusing for newcomers plus I'm not doing anything out of the world here. Its just that this is an Android-specific case where such principles don't apply. Those principles are generalised for every technology but there are instances where you have to do what you have to do.

Reply:
Well, for performance, its just the first time, not that much. But the main point is unnecessary memory occupation which will last application lifecycle.

Reply:
Since retrofit uses okHttp, here a quote from them:
> OkHttp performs best when you create a single OkHttpClient instance and reuse it for all of your HTTP calls. This is because each client holds its own connection pool and thread pools. Reusing connections and threads reduces latency and saves memory. Conversely, creating a client for each request wastes resources on idle pools.

Reply:
We can create a singleton of OkHttpClient to create multiple Retrofit instances. Just pass in the already initialized client within the `Retrofit.Builder().baseUrl("abc").client(okHttpClient)`

Reply:
> Here's the LB server [file](https://github.com/metabrainz/listenbrainz-server/blob/master/frontend/js/src/utils/APIService.ts) I'm talking about.

This is really bad code. I might talk with the team about it soon enough. Don't reference the LB code too much because it is really not the best. A lot of things are hacks.